<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Example Notebook Running SPARCSpy on datasets with multiple images &mdash; SPARCSpy  documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/fonts.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Example Notebook showcasing the different segmentation workflows" href="example_notebook_segmentation_workflows.html" />
    <link rel="prev" title="Example Notebooks" href="../example_notebooks.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> SPARCSpy
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">SPARCSpy Ecosystem</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/installation.html">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quickstart/computational_workflow.html">SPARCSpy workflow</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../quickstart/computational_workflow.html#segmentation">Segmentation</a></li>
<li class="toctree-l3"><a class="reference internal" href="../quickstart/computational_workflow.html#extraction">Extraction</a></li>
<li class="toctree-l3"><a class="reference internal" href="../quickstart/computational_workflow.html#classification">Classification</a></li>
<li class="toctree-l3"><a class="reference internal" href="../quickstart/computational_workflow.html#selection">Selection</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="example_sparcspy_project.html">A walk through the SPARCSpy Ecosystem</a><ul>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Import-Required-Libraries">Import Required Libraries</a></li>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Generate-the-Project-Structure">Generate the Project Structure</a></li>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Load-Imaging-Data">Load Imaging Data</a><ul>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Method-1:-loading-images-from-file">Method 1: loading images from file</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Method-2:-loading-images-from-numpy-array">Method 2: loading images from numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Looking-at-the-loaded-images">Looking at the loaded images</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Generating-a-Segmentation">Generating a Segmentation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Looking-at-Segmentation-Results">Looking at Segmentation Results</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Extracting-single-cell-images">Extracting single-cell images</a><ul>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Look-at-extracted-single-cell-images">Look at extracted single-cell images</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Classification-of-extracted-single-cells">Classification of extracted single-cells</a><ul>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#looking-at-the-generated-results">looking at the generated results</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html#Exporting-Cutting-countours-for-excision-on-the-LMD7">Exporting Cutting countours for excision on the LMD7</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../pipeline.html">SPARCSpy Workflow</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/project.html">SPARCSpy Projects</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/project.html#project-classes">Project classes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/project.html#project">1. Project</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/project.html#timecourseproject">2. TimecourseProject</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/project.html#initializing-a-sparcspy-project">Initializing a SPARCSpy Project</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/project.html#loading-images-into-a-sparcspy-project">Loading Images into a SPARCSpy project</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/project.html#calling-the-sparcspy-workflow">Calling the SPARCSpy workflow</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/project.html#rerunning-specific-steps">Rerunning specific steps</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/config.html">Config Files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/segmentation.html">Segmentation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/segmentation.html#segmentation-classes">Segmentation classes</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#id2">1. Segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#shardedsegmentation">2. ShardedSegmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#timecoursesegmentation">3. TimecourseSegmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#multithreadedsegmentation">4. MultithreadedSegmentation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/segmentation.html#segmentation-workflows">Segmentation Workflows</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#wga-segmentation">WGA segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#dapi-segmentation">DAPI segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#cytosol-cellpose-segmentation">Cytosol Cellpose segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../pipeline/segmentation.html#dapi-cellpose-segmentation">DAPI Cellpose segmentation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/extraction.html">Extraction</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/extraction.html#extraction-classes">Extraction classes</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pipeline/extraction.html#single-cell-image-datasets">Single-cell image datasets</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/classification.html">Classification</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pipeline/selection.html">Selection</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../example_notebooks.html">Example Notebooks</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../example_notebooks.html#basic-sparcspy-project">Basic SPARCSpy Project</a><ul>
<li class="toctree-l3"><a class="reference internal" href="example_sparcspy_project.html">A walk through the SPARCSpy Ecosystem</a><ul>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Import-Required-Libraries">Import Required Libraries</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Generate-the-Project-Structure">Generate the Project Structure</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Load-Imaging-Data">Load Imaging Data</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Generating-a-Segmentation">Generating a Segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Extracting-single-cell-images">Extracting single-cell images</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Classification-of-extracted-single-cells">Classification of extracted single-cells</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_sparcspy_project.html#Exporting-Cutting-countours-for-excision-on-the-LMD7">Exporting Cutting countours for excision on the LMD7</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2 current"><a class="reference internal" href="../example_notebooks.html#timecourse-sparcspy-project">Timecourse SPARCSpy Project</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">Example Notebook Running SPARCSpy on datasets with multiple images</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#Some-Helper-Functions-for-visualization-of-results">Some Helper Functions for visualization of results</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Initialize-a-Project">Initialize a Project</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Input-Data-Format">Input Data Format</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Example-Load-data-using-the-from-files-function">Example Load data using the from files function</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Example-loading-data-via-a-numpy-array">Example loading data via a numpy array</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Segmentation">Segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="#Single-Cell-Extraction">Single-Cell Extraction</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../example_notebooks.html#segmentation-workflows">Segmentation Workflows</a><ul>
<li class="toctree-l3"><a class="reference internal" href="example_notebook_segmentation_workflows.html">Example Notebook showcasing the different segmentation workflows</a><ul>
<li class="toctree-l4"><a class="reference internal" href="example_notebook_segmentation_workflows.html#WGA-Segmentation">WGA Segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_notebook_segmentation_workflows.html#DAPI-Segmentation">DAPI Segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_notebook_segmentation_workflows.html#Cytosol-Segmentation-Cellpose">Cytosol Segmentation Cellpose</a></li>
<li class="toctree-l4"><a class="reference internal" href="example_notebook_segmentation_workflows.html#DAPI-Segmentation-Cellpose">DAPI Segmentation Cellpose</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Module API</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../sparcscmd.html">Command Line Tools</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../sparcscmd/sparcs_split.html">sparcs-split</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_split.html#positional-arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_split.html#named-arguments">Named Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_split.html#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../sparcscmd/sparcs_stat.html">sparcs-stat</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_stat.html#positional-arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_stat.html#named-arguments">Named Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_stat.html#examples">Examples</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../sparcscmd/sparcs_cleanup.html">sparcs-cleanup</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_cleanup.html#positional-arguments">Positional Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_cleanup.html#named-arguments">Named Arguments</a></li>
<li class="toctree-l3"><a class="reference internal" href="../sparcscmd/sparcs_cleanup.html#examples">Examples</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../module.html">Module Documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../module/processing.html">processing</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../module/processing.html#module-sparcscore.processing.utils">utils</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/processing.html#module-sparcscore.processing.preprocessing">preprocessing</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/processing.html#module-sparcscore.processing.segmentation">segmentation</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../module/pipeline.html">pipeline</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#base">base</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#logable">Logable</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#processingstep">ProcessingStep</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#project">project</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#id1">Project</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#timecourseproject">TimecourseProject</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#segmentation">segmentation</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#id3">Segmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#shardedsegmentation">ShardedSegmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#timecoursesegmentation">TimecourseSegmentation</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#multithreadedtimecoursesegmentation">MultithreadedTimecourseSegmentation</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#module-sparcscore.pipeline.workflows">workflows</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#extraction">extraction</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#hdf5cellextraction">HDF5CellExtraction</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#timecoursehdf5cellextraction">TimecourseHDF5CellExtraction</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#classification">classification</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#mlclusterclassifier">MLClusterClassifier</a></li>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#cellfeaturizer">CellFeaturizer</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../module/pipeline.html#selection">selection</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../module/pipeline.html#lmdselection">LMDSelection</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../module/ml.html">ml</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#module-sparcscore.ml.datasets">datasets</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#module-sparcscore.ml.metrics">metrics</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#models">models</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#plmodels">plmodels</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#module-sparcscore.ml.pretrained_models">pretrained_models</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#module-sparcscore.ml.transforms">transforms</a></li>
<li class="toctree-l3"><a class="reference internal" href="../module/ml.html#module-sparcscore.ml.utils">utils</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">SPARCSpy</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a></li>
          <li class="breadcrumb-item"><a href="../example_notebooks.html">Example Notebooks</a></li>
      <li class="breadcrumb-item active">Example Notebook Running SPARCSpy on datasets with multiple images</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../_sources/pages/notebooks/Example_Notebook_TimeCourse_Project.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Example-Notebook-Running-SPARCSpy-on-datasets-with-multiple-images">
<h1>Example Notebook Running SPARCSpy on datasets with multiple images<a class="headerlink" href="#Example-Notebook-Running-SPARCSpy-on-datasets-with-multiple-images" title="Permalink to this headline"></a></h1>
<p>The SPARCSpy base Project class is focused on processing whole slide images. In case you want to process Image Datasets with multiple-images then you will need to use the SPARCSpy TimecourseProject class and associated methods. This tutorial will walk you through this process.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">from</span> <span class="nn">tqdm.notebook</span> <span class="kn">import</span> <span class="n">tqdm</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">h5py</span>

<span class="kn">from</span> <span class="nn">sparcscore.pipeline.project</span> <span class="kn">import</span> <span class="n">TimecourseProject</span>
<span class="kn">from</span> <span class="nn">sparcscore.pipeline.workflows</span> <span class="kn">import</span> <span class="n">Multithreaded_Cytosol_Cellpose_TimecourseSegmentation</span>
<span class="kn">from</span> <span class="nn">sparcscore.pipeline.extraction</span> <span class="kn">import</span> <span class="n">TimecourseHDF5CellExtraction</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/fs/home/maedler/.local/miniconda3/envs/SPARCS_dev/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<section id="Some-Helper-Functions-for-visualization-of-results">
<h2>Some Helper Functions for visualization of results<a class="headerlink" href="#Some-Helper-Functions-for-visualization-of-results" title="Permalink to this headline"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">sparcscore.processing.preprocessing</span> <span class="kn">import</span> <span class="n">percentile_normalization</span>
<span class="k">def</span> <span class="nf">plot_seg_overlay_timecourse</span><span class="p">(</span><span class="n">project_path</span><span class="p">,</span> <span class="n">channel_to_show</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">seg_mask_to_show</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">selection</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span> <span class="n">return_fig</span> <span class="o">=</span> <span class="kc">False</span><span class="p">):</span>

    <span class="n">seg_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project_path</span><span class="si">}</span><span class="s2">/segmentation/input_segmentation.h5&quot;</span>
    <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">seg_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">hf</span><span class="p">:</span>
        <span class="n">segmentation</span> <span class="o">=</span> <span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;segmentation&quot;</span><span class="p">)</span>
        <span class="n">channels</span> <span class="o">=</span> <span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;input_images&quot;</span><span class="p">)</span>

        <span class="nb">print</span><span class="p">(</span><span class="n">segmentation</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">channels</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">selection</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">segmentation</span> <span class="o">=</span> <span class="n">segmentation</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">seg_mask_to_show</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>
            <span class="n">image</span> <span class="o">=</span> <span class="n">channels</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="n">channel_to_show</span><span class="p">,</span> <span class="p">:,</span> <span class="p">:]</span>

            <span class="n">fig</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">figure</span><span class="p">()</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">percentile_normalization</span><span class="p">(</span><span class="n">image</span><span class="p">),</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;Greys_r&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">segmentation</span><span class="p">,</span> <span class="n">alpha</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span> <span class="n">cmap</span> <span class="o">=</span> <span class="s2">&quot;jet&quot;</span><span class="p">)</span>
            <span class="n">plt</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">return_fig</span><span class="p">:</span>
            <span class="k">return</span><span class="p">(</span><span class="n">fig</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">visualize_single_cells</span><span class="p">(</span><span class="n">project_path</span><span class="p">,</span> <span class="n">n_cells</span> <span class="o">=</span> <span class="mi">10</span><span class="p">):</span>
    <span class="n">cells_path</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project_path</span><span class="si">}</span><span class="s2">/extraction/data/single_cells.h5&quot;</span>

    <span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="n">cells_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">hf</span><span class="p">:</span>
        <span class="n">cells</span> <span class="o">=</span> <span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;single_cell_data&quot;</span><span class="p">)</span>
        <span class="n">n_channels</span> <span class="o">=</span> <span class="n">cells</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">n_cells</span><span class="p">,</span> <span class="n">n_channels</span><span class="p">,</span> <span class="n">figsize</span> <span class="o">=</span> <span class="p">(</span><span class="n">n_channels</span><span class="o">*</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_cells</span><span class="o">*</span><span class="mi">1</span><span class="p">))</span>

        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_cells</span><span class="p">):</span>
            <span class="n">image</span> <span class="o">=</span> <span class="n">cells</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">n</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_channels</span><span class="p">):</span>
                <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">image</span><span class="p">[</span><span class="n">n</span><span class="p">])</span>
                <span class="n">axs</span><span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="n">n</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Initialize-a-Project">
<h2>Initialize a Project<a class="headerlink" href="#Initialize-a-Project" title="Permalink to this headline"></a></h2>
<p>SPARCSpy works with a project structure that is kept the same across different projects. Each project contains all of the results from one run. Each Project has the same general structure:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>.
├── classification
│   └── classifier_name
│       └── processing.log
├── config.yml
├── extraction
├── segmentation
└── processing.log
</pre></div>
</div>
<p>At the beginning of a SPARCSpy analysis a new project is generated where configuration parameters are loaded from the config file. This determines how each processing step will be executed.Here is a minimal example including the different cellpose segmentation methods for TimecourseProjects.</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>---
name: &quot;HPA segmentation&quot;
input_channels: 3
output_channels: 5 #always add 2 to the number of channels
Multithreaded_Cytosol_Cellpose_TimecourseSegmentation:
    input_channels: 3
    output_masks: 2
    threads: 2 #the segmentation model is loaded for each segmentation process utilizing GPU memory. Depending on GPU size this needs to be adjusted
    cache: &quot;.&quot; #path to where intermediate results are written to memory mapped temp arrays
    lower_quantile_normalization:   0.000
    upper_quantile_normalization:   1.000
    median_filter_size:   4
    nucleus_segmentation:
        model: &quot;nuclei&quot;
        diameter: 400 #if this parameter is not set then cellpose will determine the diameter of the segemented shapes automatically
    cytosol_segmentation:
        model: &quot;cyto2&quot;
        diameter: 700
    chunk_size: 50
    filtering_threshold: 0.7
Multithreaded_Cytosol_Cellpose_Downsampling_TimecourseSegmentation:
    input_channels: 3
    output_masks: 2
    threads: 2
    cache: &quot;.&quot;
    lower_quantile_normalization:   0.000
    upper_quantile_normalization:   1.000
    median_filter_size:   4
    nucleus_segmentation:
        model: &quot;nuclei&quot;
    cytosol_segmentation:
        model: &quot;cyto2&quot;
    chunk_size: 50
    filtering_threshold: 0.7
    downsampling_factor: 10 #downsamples images before performing segmentation, the segmented masks are upscaled again to the original dimension
    smoothing_kernel_size: 7
Cytosol_Cellpose_Downsampling_TimecourseSegmentation:
    input_channels: 3
    output_masks: 2
    threads: 8
    cache: &quot;.&quot;
    lower_quantile_normalization:   0.000
    upper_quantile_normalization:   1.000
    median_filter_size:   4
    nucleus_segmentation:
        model: &quot;nuclei&quot;
    cytosol_segmentation:
        model: &quot;cyto2&quot;
    chunk_size: 50
    filtering_threshold: 0.7
    downsampling_factor: 10
    smoothing_kernel_size: 7
TimecourseHDF5CellExtraction:
    compression: True
    threads: 8 # threads used in multithreading
    image_size: 512 # image size in pixel
    cache: &quot;.&quot;
    hdf5_rdcc_nbytes: 5242880000 # 5gb 2048 * 2048 * 1250
    hdf5_rdcc_w0: 1
    hdf5_rdcc_nslots: 1250
</pre></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#initialize project</span>
<span class="n">project_location</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;/fs/pool/pool-mann-maedler-shared/sophia_workspace/Projects/timecourse_test_project&quot;</span>

<span class="n">project</span> <span class="o">=</span> <span class="n">TimecourseProject</span><span class="p">(</span><span class="n">project_location</span><span class="p">,</span>
                             <span class="n">config_path</span><span class="o">=</span> <span class="s2">&quot;config.yml&quot;</span><span class="p">,</span>
                             <span class="n">segmentation_f</span><span class="o">=</span><span class="n">Multithreaded_Cytosol_Cellpose_TimecourseSegmentation</span><span class="p">,</span>
                             <span class="n">extraction_f</span> <span class="o">=</span> <span class="n">TimecourseHDF5CellExtraction</span><span class="p">,</span>
                             <span class="n">overwrite</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                             <span class="n">debug</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
modifying config
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/fs/gpfs41/lv07/fileset03/home/b_mann/maedler/Software/SPARCS_dev/SPARCSpy/src/sparcscore/pipeline/project.py:113: UserWarning: There is already a directory in the location path
  warnings.warn(&#34;There is already a directory in the location path&#34;)
</pre></div></div>
</div>
</section>
<section id="Input-Data-Format">
<h2>Input Data Format<a class="headerlink" href="#Input-Data-Format" title="Permalink to this headline"></a></h2>
<p>SPARCSpy was optimized to work with Data generated using an Opera Phenix Microscope. Each Project class has several data_loader functions associated with it that can be used to load data from different sources. You can find the code under <code class="docutils literal notranslate"><span class="pre">src/sparcscore/project.py</span></code>. The Dataloaders that load images from files are optimzied for file_name notation generated by the Phenix and the associated processing scripts.</p>
<p>Data Format for <code class="docutils literal notranslate"><span class="pre">load_input_from_files</span></code>:</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>.... main directory
    - Row02_Well02_Row02_Well02
        - Timepoint001_Row02_Well02_Alexa488_zstack001Row02_Well02.tif
        - Timepoint001_Row02_Well02_mCherry_zstack001_r000_c000.tif
        - Timepoint001_Row02_Well02_DAPI_zstack001_r000_c000.tif
    - ...
</pre></div>
</div>
<p>Alternatively you can load data from a numpy array using <code class="docutils literal notranslate"><span class="pre">load_input_from_array</span></code> which can be used more flexibly to write your own dataloader:</p>
<ul class="simple">
<li><p>images needs to be supplied in the shape (N, C, X, Y) to the parameter img</p></li>
<li><p>labels need to be supplied as a tidy dataframe to the parameter label. The first two columns of the dataframe need to contian the following information: row_index, unique Image ID. Then you can append as many additional labels as you want.</p></li>
</ul>
<p>Also feel free to submit a PR with a new dataloader for a common microscopy file format.</p>
</section>
<section id="Example-Load-data-using-the-from-files-function">
<h2>Example Load data using the from files function<a class="headerlink" href="#Example-Load-data-using-the-from-files-function" title="Permalink to this headline"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#%%timeit -n 1</span>
<span class="n">channels</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;DAPI&quot;</span><span class="p">,</span> <span class="s2">&quot;Alexa488&quot;</span><span class="p">,</span> <span class="s2">&quot;mCherry&quot;</span><span class="p">]</span>
<span class="n">timepoints</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Timepoint&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">zfill</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">))]</span>
<span class="n">input_dir</span> <span class="o">=</span> <span class="s2">&quot;/fs/pool/pool-mann-maedler-shared/Test_Datasets/timecourse_small/timecourse_sorted&quot;</span>
<span class="n">plate_layout</span> <span class="o">=</span> <span class="s2">&quot;/fs/pool/pool-mann-maedler-shared/Test_Datasets/timecourse_small/plate_layout.tsv&quot;</span>

<span class="n">project</span><span class="o">.</span><span class="n">load_input_from_files</span><span class="p">(</span><span class="n">input_dir</span> <span class="o">=</span> <span class="n">input_dir</span><span class="p">,</span>  <span class="n">channels</span> <span class="o">=</span> <span class="n">channels</span><span class="p">,</span>  <span class="n">timepoints</span> <span class="o">=</span> <span class="n">timepoints</span><span class="p">,</span> <span class="n">plate_layout</span> <span class="o">=</span> <span class="n">plate_layout</span><span class="p">,</span> <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
100%|██████████| 2/2 [00:00&lt;00:00,  2.61it/s]
</pre></div></div>
</div>
<p>Images and labels are written to the <code class="docutils literal notranslate"><span class="pre">input_segmentation.h5</span></code> HDF5 file contained under segmentation in the project folder. The input images are written to the container “input_images” and the labels to “labels”. Column names are saved in a seperate dataset with the name “label_names”.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project_location</span><span class="si">}</span><span class="s2">/segmentation/input_segmentation.h5&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">hf</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;input_images&quot;</span><span class="p">))</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;labels&quot;</span><span class="p">))</span>

    <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Example Label Dataset&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;label_names&quot;</span><span class="p">)[:])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;labels&quot;</span><span class="p">)[:])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;HDF5 dataset &#34;input_images&#34;: shape (4, 3, 1080, 1080), type &#34;&lt;u2&#34;&gt;
&lt;HDF5 dataset &#34;labels&#34;: shape (4, 8), type &#34;|O&#34;&gt;
Example Label Dataset
[b&#39;index&#39; b&#39;ID&#39; b&#39;location&#39; b&#39;timepoint&#39; b&#39;well&#39; b&#39;region&#39; b&#39;Genotype&#39;
 b&#39;Treatment&#39;]
[[b&#39;0&#39; b&#39;Row02_Well04_r004_c004_Timepoint001&#39; b&#39;Row02_Well04_r004_c004&#39;
  b&#39;Timepoint001&#39; b&#39;Row02_Well04&#39; b&#39;r004_c004&#39; b&#39;WT_1&#39; b&#39;Unstimulated&#39;]
 [b&#39;1&#39; b&#39;Row02_Well04_r004_c004_Timepoint002&#39; b&#39;Row02_Well04_r004_c004&#39;
  b&#39;Timepoint002&#39; b&#39;Row02_Well04&#39; b&#39;r004_c004&#39; b&#39;WT_1&#39; b&#39;Unstimulated&#39;]
 [b&#39;2&#39; b&#39;Row02_Well02_r000_c000_Timepoint001&#39; b&#39;Row02_Well02_r000_c000&#39;
  b&#39;Timepoint001&#39; b&#39;Row02_Well02&#39; b&#39;r000_c000&#39; b&#39;WT_1&#39; b&#39;Unstimulated&#39;]
 [b&#39;3&#39; b&#39;Row02_Well02_r000_c000_Timepoint002&#39; b&#39;Row02_Well02_r000_c000&#39;
  b&#39;Timepoint002&#39; b&#39;Row02_Well02&#39; b&#39;r000_c000&#39; b&#39;WT_1&#39; b&#39;Unstimulated&#39;]]
</pre></div></div>
</div>
</section>
<section id="Example-loading-data-via-a-numpy-array">
<h2>Example loading data via a numpy array<a class="headerlink" href="#Example-loading-data-via-a-numpy-array" title="Permalink to this headline"></a></h2>
<p>the timecourse SPARCSpy project expects images of the following format: NCYX.</p>
<p>All of the images need to have the same dimensions and the same number of channels. The channels should be in the following order: nucleus, cytosol, other channels.</p>
<p>The images need to have the dtype uint16.</p>
<p>Here we have example code which reads RGB images (saved as PNGs), converts them to the expected format and loads them into a SPARCSpy timecourse project.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#read metadata of images</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;metadata.csv&quot;</span><span class="p">,</span> <span class="n">index_col</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">metadata</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>gene_id</th>
      <th>antibody</th>
      <th>species</th>
      <th>cell_line</th>
      <th>organ</th>
      <th>cellosaurusID</th>
      <th>location</th>
      <th>image_URL</th>
      <th>image_location</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>ENSG00000000003</td>
      <td>HPA004109</td>
      <td>human</td>
      <td>U2OS</td>
      <td>Mesenchymal</td>
      <td>CVCL_0042</td>
      <td>cytosol</td>
      <td>http://images.proteinatlas.org/4109/23_H11_1_b...</td>
      <td>images/00005.jpg</td>
    </tr>
    <tr>
      <th>1</th>
      <td>ENSG00000000003</td>
      <td>HPA004109</td>
      <td>human</td>
      <td>U2OS</td>
      <td>Mesenchymal</td>
      <td>CVCL_0042</td>
      <td>cytosol</td>
      <td>http://images.proteinatlas.org/4109/23_H11_2_b...</td>
      <td>images/00006.jpg</td>
    </tr>
    <tr>
      <th>2</th>
      <td>ENSG00000000460</td>
      <td>HPA024451</td>
      <td>human</td>
      <td>U2OS</td>
      <td>Mesenchymal</td>
      <td>CVCL_0042</td>
      <td>NaN</td>
      <td>http://images.proteinatlas.org/24451/226_G2_1_...</td>
      <td>images/00017.jpg</td>
    </tr>
    <tr>
      <th>3</th>
      <td>ENSG00000000460</td>
      <td>HPA024451</td>
      <td>human</td>
      <td>U2OS</td>
      <td>Mesenchymal</td>
      <td>CVCL_0042</td>
      <td>NaN</td>
      <td>http://images.proteinatlas.org/24451/226_G2_2_...</td>
      <td>images/00018.jpg</td>
    </tr>
    <tr>
      <th>4</th>
      <td>ENSG00000000938</td>
      <td>HPA002024</td>
      <td>human</td>
      <td>U2OS</td>
      <td>Mesenchymal</td>
      <td>CVCL_0042</td>
      <td>plasma membrane</td>
      <td>http://images.proteinatlas.org/2024/59_H5_1_bl...</td>
      <td>images/00023.jpg</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#read images</span>

<span class="kn">from</span> <span class="nn">multiprocessing</span> <span class="kn">import</span> <span class="n">Pool</span>
<span class="kn">from</span> <span class="nn">skimage.transform</span> <span class="kn">import</span> <span class="n">resize</span>

<span class="k">def</span> <span class="nf">read_image</span><span class="p">(</span><span class="n">img</span><span class="p">):</span>
    <span class="n">I</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">imread</span><span class="p">(</span><span class="n">img</span><span class="p">)</span>
    <span class="c1"># convert from rgb to brg (nucleus, cytosol, other)</span>
    <span class="n">I</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">roll</span><span class="p">(</span><span class="n">I</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">I</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">!=</span> <span class="mi">2048</span><span class="p">:</span>
        <span class="n">I</span> <span class="o">=</span> <span class="n">resize</span><span class="p">(</span><span class="n">I</span><span class="p">,</span> <span class="p">(</span><span class="mi">2048</span><span class="p">,</span> <span class="mi">2048</span><span class="p">))</span> <span class="c1">#some of the images have different shapes so we force resize of those images that dont match the most common shape</span>

    <span class="c1"># convert from H x W x C to C x H x W</span>
    <span class="n">I</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">moveaxis</span><span class="p">(</span><span class="n">I</span><span class="p">,</span> <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">0</span><span class="p">,</span>  <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
    <span class="n">I</span> <span class="o">=</span> <span class="n">I</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;float64&quot;</span><span class="p">)</span>
    <span class="n">I</span> <span class="o">=</span> <span class="p">(</span><span class="n">I</span><span class="o">/</span><span class="n">I</span><span class="o">.</span><span class="n">max</span><span class="p">()</span><span class="o">*</span><span class="mi">65535</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="s2">&quot;uint16&quot;</span><span class="p">)</span>  <span class="c1">#normalize all images to the same range and then convert to uint16</span>
    <span class="k">return</span> <span class="n">I</span>

<span class="n">image_paths</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">[</span><span class="s2">&quot;image_location&quot;</span><span class="p">]</span>

<span class="n">pool</span> <span class="o">=</span> <span class="n">Pool</span><span class="p">(</span><span class="mi">12</span><span class="p">)</span> <span class="c1">#allow reading images over multiple threads</span>
<span class="n">images</span> <span class="o">=</span> <span class="n">pool</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">read_image</span><span class="p">,</span> <span class="n">image_paths</span><span class="p">)</span>
<span class="n">images</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">stack</span><span class="p">(</span><span class="n">images</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>The images have the expected shape (10 images, 3 channels, 2048px, 2048px) and look as expected</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="n">images</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="n">images</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>

<span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">images</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">3</span><span class="p">]:</span>
    <span class="n">fig</span><span class="p">,</span> <span class="n">axs</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">imshow</span><span class="p">(</span><span class="n">i</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span>
    <span class="n">axs</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s2">&quot;off&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(10, 3, 2048, 2048)
uint16
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_15_1.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_15_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_15_2.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_15_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_15_3.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_15_3.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#format the metadata to the expected format</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">reset_index</span><span class="p">()</span>
<span class="n">metadata</span> <span class="o">=</span> <span class="n">metadata</span><span class="o">.</span><span class="n">get</span><span class="p">([</span><span class="s2">&quot;index&quot;</span><span class="p">,</span> <span class="s2">&quot;image_URL&quot;</span><span class="p">,</span> <span class="s1">&#39;gene_id&#39;</span><span class="p">,</span> <span class="s1">&#39;antibody&#39;</span><span class="p">,</span> <span class="s1">&#39;species&#39;</span><span class="p">,</span> <span class="s1">&#39;cell_line&#39;</span><span class="p">,</span> <span class="s1">&#39;organ&#39;</span><span class="p">,</span> <span class="s1">&#39;cellosaurusID&#39;</span><span class="p">])</span> <span class="c1">#here we use the image_URL as a unqiue image id</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">#load data into sparcspy</span>
<span class="n">project</span><span class="o">.</span><span class="n">load_input_from_array</span><span class="p">(</span><span class="n">images</span><span class="p">,</span>
                             <span class="n">label</span> <span class="o">=</span> <span class="n">metadata</span><span class="p">,</span>
                             <span class="n">overwrite</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;KeysViewHDF5 [&#39;input_images&#39;, &#39;label_names&#39;, &#39;labels&#39;]&gt;
</pre></div></div>
</div>
</section>
<section id="Segmentation">
<h2>Segmentation<a class="headerlink" href="#Segmentation" title="Permalink to this headline"></a></h2>
<p>SPARCSpy has different segmentation workflows between which you can choose. If you run a timecourse project you will need to also select a timecourse segementation method.</p>
<p><strong>notes on cellpose segmentation methods:</strong></p>
<p>Cellpose currently does not allow for the processing of image batches. Each image will thus be segmented individually on the GPU. In case you have large images this isnt an issue since the GPU is still fully utilzed (the sharding parameter can be used to set an optimal shard size for maximizing GPU utilization). For a large quantity of small images this leads to a very inefficient processing. To improve the speed a bit you can try running with multiple threads. Each thread will also load the
model though so it requires quite some Memory overhead. This is hopefully an issue we can address in future releases. If its something you would like to work on get in touch with the developers we are happy for help!</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">project</span><span class="o">.</span><span class="n">segment</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:root:Temp mmap arrays are written to /tmp/temp_mmap_t7889uf5. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:Folder /tmp/temp_mmap_t7889uf5 with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmappc5_15oq. Cleanup of this folder is OS dependant, and might need to be triggered manually!
Generating Shards: 100%|██████████| 10/10 [00:00&lt;00:00, 13025.79it/s]
  0%|          | 0/10 [00:00&lt;?, ?it/s]WARNING:root:Folder /tmp/./temp_mmappc5_15oq with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapry9pprzj. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:Folder /tmp/./temp_mmappc5_15oq with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapmnfdajf6. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:Folder /tmp/./temp_mmappc5_15oq with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:Folder /tmp/./temp_mmappc5_15oq with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapbc_n4alu. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmap34detioy. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:Folder /tmp/./temp_mmap34detioy with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapd4b8tluv. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:Folder /tmp/./temp_mmapmnfdajf6 with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapp3hs2u_r. Cleanup of this folder is OS dependant, and might need to be triggered manually!
WARNING:root:Folder /tmp/./temp_mmapry9pprzj with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapclg9qjls. Cleanup of this folder is OS dependant, and might need to be triggered manually!
 10%|█         | 1/10 [01:22&lt;12:22, 82.49s/it]WARNING:root:Folder /tmp/./temp_mmapbc_n4alu with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapso3tyehs. Cleanup of this folder is OS dependant, and might need to be triggered manually!
 50%|█████     | 5/10 [01:35&lt;01:15, 15.15s/it]WARNING:root:Folder /tmp/./temp_mmapd4b8tluv with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapa3dk_m7q. Cleanup of this folder is OS dependant, and might need to be triggered manually!
 60%|██████    | 6/10 [01:43&lt;00:53, 13.43s/it]WARNING:root:Folder /tmp/./temp_mmapp3hs2u_r with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmapyaav5jws. Cleanup of this folder is OS dependant, and might need to be triggered manually!
100%|██████████| 10/10 [02:32&lt;00:00, 15.24s/it]
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
All segmentations are done.
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
Adjusting Indexes: 100%|██████████| 10/10 [00:11&lt;00:00,  1.14s/it]
</pre></div></div>
</div>
<p>Segmentation results are written to the <code class="docutils literal notranslate"><span class="pre">input_segmentation.h5</span></code> HDF5 file in a new data container called “segmentation”. In addition a csv file called classes.csv is generated which contains all the cell_ids that are not located in image edges and will be further processed.</p>
<p>If we look at the segmentation dataset we can see that it contains a numpy array containing two segmentation masks for each input image: the nuclear segmentation and the cytosol segmentation generated by our chosen segmentation method. If you use the method CytosolOnly both of the masks will be identical.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_seg_overlay_timecourse</span><span class="p">(</span><span class="n">project_location</span><span class="p">,</span> <span class="n">channel_to_show</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">plot_seg_overlay_timecourse</span><span class="p">(</span><span class="n">project_location</span><span class="p">,</span> <span class="n">channel_to_show</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">plot_seg_overlay_timecourse</span><span class="p">(</span><span class="n">project_location</span><span class="p">,</span> <span class="n">channel_to_show</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">seg_mask_to_show</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
<span class="n">plot_seg_overlay_timecourse</span><span class="p">(</span><span class="n">project_location</span><span class="p">,</span> <span class="n">channel_to_show</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">seg_mask_to_show</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;HDF5 dataset &#34;segmentation&#34;: shape (10, 2, 2048, 2048), type &#34;&lt;u4&#34;&gt;
&lt;HDF5 dataset &#34;input_images&#34;: shape (10, 3, 2048, 2048), type &#34;&lt;u2&#34;&gt;
&lt;HDF5 dataset &#34;segmentation&#34;: shape (10, 2, 2048, 2048), type &#34;&lt;u4&#34;&gt;
&lt;HDF5 dataset &#34;input_images&#34;: shape (10, 3, 2048, 2048), type &#34;&lt;u2&#34;&gt;
&lt;HDF5 dataset &#34;segmentation&#34;: shape (10, 2, 2048, 2048), type &#34;&lt;u4&#34;&gt;
&lt;HDF5 dataset &#34;input_images&#34;: shape (10, 3, 2048, 2048), type &#34;&lt;u2&#34;&gt;
&lt;HDF5 dataset &#34;segmentation&#34;: shape (10, 2, 2048, 2048), type &#34;&lt;u4&#34;&gt;
&lt;HDF5 dataset &#34;input_images&#34;: shape (10, 3, 2048, 2048), type &#34;&lt;u2&#34;&gt;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_1.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_2.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_2.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_3.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_3.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_4.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_21_4.png" />
</div>
</div>
</section>
<section id="Single-Cell-Extraction">
<h2>Single-Cell Extraction<a class="headerlink" href="#Single-Cell-Extraction" title="Permalink to this headline"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">project</span><span class="o">.</span><span class="n">extract</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:root:Folder /tmp/./temp_mmappc5_15oq with temp mmap arrays is being deleted.All existing temp mmapp arrays will be unusable!
WARNING:root:New temp folder location. Temp mmap arrays are written to /tmp/./temp_mmape8paw22r. Cleanup of this folder is OS dependant, and might need to be triggered manually!
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(80, 5, 512, 512)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
 10%|█         | 1/10 [00:04&lt;00:37,  4.19s/it]/fs/gpfs41/lv07/fileset03/home/b_mann/maedler/Software/SPARCS_dev/SPARCSpy/src/sparcscore/pipeline/extraction.py:877: RuntimeWarning: invalid value encountered in cast
  px_centers = np.round(center_nuclei).astype(int)
100%|██████████| 10/10 [00:17&lt;00:00,  1.71s/it]
</pre></div></div>
</div>
<p>The extracted single-cell images are written to a h5py file single_cells.h5 located under <code class="docutils literal notranslate"><span class="pre">extraction\data</span></code> within the project folder.</p>
<p>The file contains four datasets: <code class="docutils literal notranslate"><span class="pre">single_cell_data</span></code>, <code class="docutils literal notranslate"><span class="pre">single_cell_index</span></code>, <code class="docutils literal notranslate"><span class="pre">single_cell_index_labelled</span></code>, <code class="docutils literal notranslate"><span class="pre">label_names</span></code>.</p>
<p><code class="docutils literal notranslate"><span class="pre">single_cell_data</span></code> contains the extracted single cell images while <code class="docutils literal notranslate"><span class="pre">single_cell_index</span></code> contains the cell id of the extracted cell at that location.</p>
<p>The other two datacontainers integrated the labels from the base dataset and attach them to each cell.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project_location</span><span class="si">}</span><span class="s2">/extraction/data/single_cells.h5&quot;</span><span class="p">)</span><span class="k">as</span> <span class="n">hf</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;label_names&quot;</span><span class="p">)[:])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;single_cell_index_labelled&quot;</span><span class="p">)[</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;KeysViewHDF5 [&#39;label_names&#39;, &#39;single_cell_data&#39;, &#39;single_cell_index&#39;, &#39;single_cell_index_labelled&#39;]&gt;
[b&#39;index&#39; b&#39;cellid&#39; b&#39;image_URL&#39; b&#39;gene_id&#39; b&#39;antibody&#39; b&#39;species&#39;
 b&#39;cell_line&#39; b&#39;organ&#39; b&#39;cellosaurusID&#39;]
[b&#39;2&#39; b&#39;4&#39;
 b&#39;http://images.proteinatlas.org/4109/23_H11_1_blue_red_green.jpg&#39;
 b&#39;ENSG00000000003&#39; b&#39;HPA004109&#39; b&#39;human&#39; b&#39;U2OS&#39; b&#39;Mesenchymal&#39;
 b&#39;CVCL_0042&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># single cell index contains two columns, the first is the index position in the hdf5 datacontainer, the second is the cell id</span>

<span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project_location</span><span class="si">}</span><span class="s2">/extraction/data/single_cells.h5&quot;</span><span class="p">)</span><span class="k">as</span> <span class="n">hf</span><span class="p">:</span>

    <span class="c1">#print(hf[&quot;single_cell_index&quot;][0:10])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="p">[</span><span class="s2">&quot;single_cell_index&quot;</span><span class="p">][</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[[ 0  2]
 [ 1  3]
 [ 2  4]
 [ 3  8]
 [ 4 11]
 [ 5 12]
 [ 6 13]
 [ 7 14]
 [ 8 19]
 [ 9 20]]
[2 4]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># single cell index labelled contains the same information as single_cell_index</span>
<span class="c1"># in the first two columns but adds additional columns with labelling information, the column names for</span>
<span class="c1"># the labelling information is contained in label_names</span>

<span class="k">with</span> <span class="n">h5py</span><span class="o">.</span><span class="n">File</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">project_location</span><span class="si">}</span><span class="s2">/extraction/data/single_cells.h5&quot;</span><span class="p">)</span><span class="k">as</span> <span class="n">hf</span><span class="p">:</span>

    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="p">[</span><span class="s2">&quot;label_names&quot;</span><span class="p">][:])</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">hf</span><span class="p">[</span><span class="s2">&quot;single_cell_index_labelled&quot;</span><span class="p">][</span><span class="mi">2</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[b&#39;index&#39; b&#39;cellid&#39; b&#39;image_URL&#39; b&#39;gene_id&#39; b&#39;antibody&#39; b&#39;species&#39;
 b&#39;cell_line&#39; b&#39;organ&#39; b&#39;cellosaurusID&#39;]
[b&#39;2&#39; b&#39;4&#39;
 b&#39;http://images.proteinatlas.org/4109/23_H11_1_blue_red_green.jpg&#39;
 b&#39;ENSG00000000003&#39; b&#39;HPA004109&#39; b&#39;human&#39; b&#39;U2OS&#39; b&#39;Mesenchymal&#39;
 b&#39;CVCL_0042&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[39]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">visualize_single_cells</span><span class="p">(</span><span class="n">project_location</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_28_0.png" src="../../_images/pages_notebooks_Example_Notebook_TimeCourse_Project_28_0.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="../example_notebooks.html" class="btn btn-neutral float-left" title="Example Notebooks" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="example_notebook_segmentation_workflows.html" class="btn btn-neutral float-right" title="Example Notebook showcasing the different segmentation workflows" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023 Sophia Mädler, Georg Wallmann and Niklas Schmacke.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>